### PR TITLE
Optimize the binary size of the XOrg color table - part 2

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

This optimizes the binary size of the xorg color table. Think of it as a follow-up of #7929. 

 
## PR Checklist
* [ ] Closes #xxx
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [X] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Inspired by @DHowett 's suggestion in #7929, I figured why not calculating all the color variants. Turns out there's a way described in [wikipedia](https://en.wikipedia.org/wiki/X11_color_names). So I implemented it and guess what, another 1KB is stripped from the product. Besides, it makes the color table less lengthy (less scary to read).

## Validation Steps Performed

Tested passed. Manually tested.

### CODE REVIEW

File: src/types/ut_types/UtilsTests.cpp [hunk 1]
```diff
@@ -190,9 +190,9 @@ void UtilsTests::TestColorFromXTermColor()
     _VerifyXTermColorResult(L"yellow", RGB(255, 255, 0));
     _VerifyXTermColorResult(L"yellow3", RGB(205, 205, 0));
     _VerifyXTermColorResult(L"wheat", RGB(245, 222, 179));
-    _VerifyXTermColorResult(L"wheat4", RGB(139, 126, 102));
+    _VerifyXTermColorResult(L"wheat4", RGB(139, 126, 101));
```
Comment (skyline75489) [inline src/types/ut_types/UtilsTests.cpp:193]: The formula itself is pretty accurate. But the rounding method is not ideal. Do we have a rounding method that can round to the nearest integer? For example the real value here is actually ~101.8, so the result should be 102. With the correct rounding method I think the test case can remain intact.

**Comment by skyline75489** [comment]
I've gotta admit I expect a little more than 1KB. TIL a valuable lesson that integers take (surprisingly) less space than strings. I mean I know it, of course. I just didn't really *feel* it.

File: src/types/colorTable.cpp [hunk 2]
```diff
@@ -286,85 +286,123 @@ static constexpr std::array<til::color, 256> standardXterm256ColorTable{
...
+        break;
+    }
+
+    return til::color(
+        ::base::saturated_cast<uint8_t>(baseColor.r * ratio),
+        ::base::saturated_cast<uint8_t>(baseColor.g * ratio),
+        ::base::saturated_cast<uint8_t>(baseColor.b * ratio));
+}
+
+static constexpr til::presorted_static_map xorgAppStandardVariantColorTable{
```
Comment (skyline75489) [inline src/types/colorTable.cpp:324]: With all these XOrg color things, I've completely run out of ideas of a good variable name 😦

File: src/types/colorTable.cpp [hunk 3]
```diff
@@ -561,13 +600,41 @@ try
...
     {
-        const auto colors = variantColorIter->second;
-        if (variantIndex < colors.size())
+        const auto baseColor = variantColorIter->second;
+        if (variantIndex > 4)
+        {
+            return std::nullopt;
+        }
+
+        if (foundVariant)
```
Comment (skyline75489) [inline src/types/colorTable.cpp:631]: If `variantIndex` is 1 by default, this logic can be simplified. But I'm always hesitate to initialize some integer to 1 by default. It's just not as natural as 0..

File: src/types/colorTable.cpp [hunk 4]
```diff
@@ -286,85 +286,123 @@ static constexpr std::array<til::color, 256> standardXterm256ColorTable{
...
+        ratio = 0.932f;
+        break;
+    case 3:
+        ratio = 0.804f;
+        break;
+    case 4:
+        ratio = 0.548f;
+        break;
+    default:
+        break;
```
Comment (DHowett) [inline src/types/colorTable.cpp:315]: if you switch to a similar algorithm to the one we used for gray ((255*932)+500/1000), you'll get rounding behavior on truncate for free _and_ we won't need floating point math :smile:

File: src/types/colorTable.cpp [hunk 5]
```diff
@@ -286,85 +286,123 @@ static constexpr std::array<til::color, 256> standardXterm256ColorTable{
...
+
+static constexpr til::presorted_static_map xorgAppStandardVariantColorTable{
+    std::pair{ "antiquewhite"sv, std::array<til::color, 2>{ til::color{ 250, 235, 215 }, til::color{ 255, 239, 219 } } },
+    std::pair{ "brown"sv, std::array<til::color, 2>{ til::color{ 165, 42, 42 }, til::color{ 255, 64, 64 } } },
+    std::pair{ "burlywood"sv, std::array<til::color, 2>{ til::color{ 222, 184, 135 }, til::color{ 255, 211, 155 } } },
+    std::pair{ "cadetblue"sv, std::array<til::color, 2>{ til::color{ 95, 158, 160 }, til::color{ 152, 245, 255 } } },
+    std::pair{ "chocolate"sv, std::array<til::color, 2>{ til::color{ 210, 105, 30 }, til::color{ 255, 127, 36 } } },
+    std::pair{ "coral"sv, std::array<til::color, 2>{ til::color{ 255, 127, 80 }, til::color{ 255, 114, 86 } } },
+    std::pair{ "darkgoldenrod"sv, std::array<til::color, 2>{ til::color{ 184, 134, 11 }, til::color{ 255, 185, 15 } } },
+    std::pair{ "darkolivegreen"sv, std::array<til::color, 2>{ til::color{ 85, 107, 47 } } },
```
Comment (Lo0oG) [inline src/types/colorTable.cpp:332]: In the xorgAppStandardVariantColorTable table, why do darkolivegreen and turquoise only have one item where as all the others have two?

Comment (skyline75489) [reply src/types/colorTable.cpp:332]: Sharp eyes. That is a mistake. Thanks for catching this.

Huh, I thought compiler would warn me about some things. Say I have a `std::array<x, 2>` and I put 3 objects in it, the compiler would warn me. But it does not care if the elements are not enough?

**Comment by skyline75489** [comment]
I updated the formula. It is still slightly off. But at least no float number is needed. I don’t think people would actually care very much about x11 color being slightly off. So I guess it’s ready.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Lo0oG <notifications@github.com>
发送时间: Friday, October 16, 2020 10:36:57 PM
收件人: microsoft/terminal <terminal@noreply.github.com>
抄送: Chester Liu <skyline75489@outlook.com>; Author <author@noreply.github.com>
主题: Re: [microsoft/terminal] Optimize the binary size of the XOrg color table - part 2 (#7945)


@Lo0oG commented on this pull request.

________________________________

In src/types/colorTable.cpp<https://github.com/microsoft/terminal/pull/7945#discussion_r506493133>:

> +
+    return til::color(
+        ::base::saturated_cast<uint8_t>(baseColor.r * ratio),
+        ::base::saturated_cast<uint8_t>(baseColor.g * ratio),
+        ::base::saturated_cast<uint8_t>(baseColor.b * ratio));
+}
+
+static constexpr til::presorted_static_map xorgAppStandardVariantColorTable{
+    std::pair{ "antiquewhite"sv, std::array<til::color, 2>{ til::color{ 250, 235, 215 }, til::color{ 255, 239, 219 } } },
+    std::pair{ "brown"sv, std::array<til::color, 2>{ til::color{ 165, 42, 42 }, til::color{ 255, 64, 64 } } },
+    std::pair{ "burlywood"sv, std::array<til::color, 2>{ til::color{ 222, 184, 135 }, til::color{ 255, 211, 155 } } },
+    std::pair{ "cadetblue"sv, std::array<til::color, 2>{ til::color{ 95, 158, 160 }, til::color{ 152, 245, 255 } } },
+    std::pair{ "chocolate"sv, std::array<til::color, 2>{ til::color{ 210, 105, 30 }, til::color{ 255, 127, 36 } } },
+    std::pair{ "coral"sv, std::array<til::color, 2>{ til::color{ 255, 127, 80 }, til::color{ 255, 114, 86 } } },
+    std::pair{ "darkgoldenrod"sv, std::array<til::color, 2>{ til::color{ 184, 134, 11 }, til::color{ 255, 185, 15 } } },
+    std::pair{ "darkolivegreen"sv, std::array<til::color, 2>{ til::color{ 85, 107, 47 } } },


In the xorgAppStandardVariantColorTable table, why do darkolivegreen and turquoise only have one item where as all the others have two?

―

**Review by miniksa — APPROVED** [review]
I agree that being off by 1ish in R, G, and/or B is no big deal as it shouldn't be visually apparent to anyone. Normally I wouldn't chase down one more KB, but since you laid it all out here already... (and since Dustin and I were musing about creative ways to do it for funsies yesterday afternoon...) 

Approved. And thanks.

**Review by DHowett — COMMENTED** [review]
So, I'm less convinced about this one.

I think we're at a good point in the complexity, effort and size curves. This PR introduces another table, I think, bring our total up to three.

1. Colors whose base and first variant differ
2. Colors whose base and first variant do not differ
3. Colors with no variants

I like that we can calculate the colors based on these values, but it worries me that we've gone from something largely machine-generated (name -> color) to something more manual.

We're also at a good place with regards to _perfection_. Things are 1:1 accurate for all the x11 colors... so overall I'm not too worried about the 4k we're spending.

**Comment by skyline75489** [comment]
I had the same concern as Dustin. Besides the 1kb isn’t really convincing enough for me. So feel free to close this. If you need the extra 1kb in the future, you’ll know where to squeeze it out

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Dustin L. Howett <notifications@github.com>
发送时间: Saturday, October 17, 2020 1:11:27 AM
收件人: microsoft/terminal <terminal@noreply.github.com>
抄送: Chester Liu <skyline75489@outlook.com>; Author <author@noreply.github.com>
主题: Re: [microsoft/terminal] Optimize the binary size of the XOrg color table - part 2 (#7945)


@DHowett commented on this pull request.

So, I'm less convinced about this one.

I think we're at a good point in the complexity, effort and size curves. This PR introduces another table, I think, bring our total up to three.

  1.  Colors whose base and first variant differ
  2.  Colors whose base and first variant do not differ
  3.  Colors with no variants

I like that we can calculate the colors based on these values, but it worries me that we've gone from something largely machine-generated (name -> color) to something more manual.

We're also at a good place with regards to perfection. Things are 1:1 accurate for all the x11 colors... so overall I'm not too worried about the 4k we're spending.

―

**Comment by skyline75489** [comment]
I like how bot edits my emails to make them cleaner. But apparently that does not work all the time? Or is the bot just having a bad mood.

I’d really like to see what creative ways you guys have to make the binary smaller. Sounds like a lot of fun. Reminds me of the good old days when I was trying to squeeze one percent of CPU time in the rendering process.

Anyway, feel free to ignore this. The previous PR is good enough for me.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: 刘 佳亮 <skyline75489@outlook.com>
发送时间: Saturday, October 17, 2020 5:58:28 AM
收件人: microsoft/terminal <reply@reply.github.com>; microsoft/terminal <terminal@noreply.github.com>
抄送: Author <author@noreply.github.com>
主题: Re: [microsoft/terminal] Optimize the binary size of the XOrg color table - part 2 (#7945)

I had the same concern as Dustin. Besides the 1kb isn’t really convincing enough for me. So feel free to close this. If you need the extra 1kb in the future, you’ll know where to squeeze it out

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Dustin L. Howett <notifications@github.com>
发送时间: Saturday, October 17, 2020 1:11:27 AM
收件人: microsoft/terminal <terminal@noreply.github.com>
抄送: Chester Liu <skyline75489@outlook.com>; Author <author@noreply.github.com>
主题: Re: [microsoft/terminal] Optimize the binary size of the XOrg color table - part 2 (#7945)


@DHowett commented on this pull request.

So, I'm less convinced about this one.

I think we're at a good point in the complexity, effort and size curves. This PR introduces another table, I think, bring our total up to three.

  1.  Colors whose base and first variant differ
  2.  Colors whose base and first variant do not differ
  3.  Colors with no variants

I like that we can calculate the colors based on these values, but it worries me that we've gone from something largely machine-generated (name -> color) to something more manual.

We're also at a good place with regards to perfection. Things are 1:1 accurate for all the x11 colors... so overall I'm not too worried about the 4k we're spending.

―

**Comment by miniksa** [comment]
K, well, I wasn't a hard "yes" or a hard "no" on this and Dustin seems to have slightly stronger feelings than I do. And with you backing out of it too, Chester, I think I'll just close it for now to pare down the PR list. Thanks for playing performance games with us!

**Pull request closed without merging**